### PR TITLE
Update jvmtiGetThreadState() to handle virtual thread

### LINKED ISSUE #15762: Loom: VirtualThread support for JVMTI  GetThreadState
Part of #15183 
https://download.java.net/java/early_access/jdk19/docs/specs/jvmti.html#GetThreadState

Need to check if thread object is VirtualThread, return the correct state based on VirtualThread.state

See: https://github.com/eclipse-openj9/openj9/issues/15183#issuecomment-1167767823

### PR DESCRIPTION
**Pull request by hangshao0**
1. If jvmtiGetThreadState() is called on a virtual thread, retrieve
java.lang.VirtualThread.State and map it to
JVMTI_JAVA_LANG_THREAD_STATE_XXX.

2. Fix assertion failure.

Closes #15762

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

### CODE REVIEW

**Comment by hangshao0** [comment]
FYI @fengxue-IS

File: runtime/jvmti/jvmtiHelpers.c [hunk 1]
```diff
@@ -772,6 +772,87 @@ getThreadState(J9VMThread *currentThread, j9object_t threadObject)
...
+		{
+			JNIEnv *env = (JNIEnv *)currentThread;
+			jfieldID fid = NULL;
+			jclass jlThread = NULL;
+			if (hasVMAccess) {
+				vm->internalVMFunctions->internalExitVMToJNI(currentThread);
+			}
+			jlThread = (*env)->FindClass(env, "java/lang/Thread");
+			Assert_JVMTI_notNull(jlThread);
+			fid = (*env)->GetFieldID(env, jlThread, "container", "Ljdk/internal/vm/ThreadContainer;");
```
Comment (tajila) [inline runtime/jvmti/jvmtiHelpers.c:801]: this can fail with OOM

Comment (hangshao0) [reply runtime/jvmti/jvmtiHelpers.c]: Fixed

File: runtime/jvmti/jvmtiHelpers.c [hunk 2]
```diff
@@ -772,6 +772,87 @@ getThreadState(J9VMThread *currentThread, j9object_t threadObject)
...
+		case JVMTI_VTHREAD_STATE_PARKED_SUSPENDED:
+			/* Fall Through */
+		case JVMTI_VTHREAD_STATE_PINNED:
+			rc = JVMTI_JAVA_LANG_THREAD_STATE_WAITING;
+			break;
+		case JVMTI_VTHREAD_STATE_TERMINATED:
+			rc = JVMTI_JAVA_LANG_THREAD_STATE_TERMINATED;
+			break;
+		default:
+			Assert_JVMTI_true(0);
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/jvmti/jvmtiHelpers.c]: Fixed

File: runtime/jvmti/jvmtiHelpers.c [hunk 3]
```diff
@@ -772,6 +772,87 @@ getThreadState(J9VMThread *currentThread, j9object_t threadObject)
...
+			rc = JVMTI_ERROR_INTERNAL;
+		}
+		releaseVMThread(currentThread, targetThread, thread);
+	} else {
+		/* This is unreachable. */
+		Assert_JVMTI_true(0);
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/jvmti/jvmtiHelpers.c]: Fixed

**Comment by hangshao0** [comment]
Update to map `JVMTI_VTHREAD_STATE_PARKED` to `JVMTI_THREAD_STATE_SLEEPING`. 

Currently JCL code `j.l.VirtualThread.threadState()` translates `JVMTI_VTHREAD_STATE_PARKED` to `JVMTI_JAVA_LANG_THREAD_STATE_WAITING`. However a jvmti test expects `JVMTI_THREAD_STATE_SLEEPING` to be set in this case.

**Comment by hangshao0** [comment]
However, the doc https://download.java.net/java/early_access/jdk19/docs/specs/jvmti.html#GetThreadState says:

`JVMTI_THREAD_STATE_PARKED ...... A virtual thread that is sleeping, in Thread.sleep, may have this state flag set instead of JVMTI_THREAD_STATE_SLEEPING. `

Not sure if the OpenJDK test will be updated.

**Comment by hangshao0** [comment]
I've reverted the change described in https://github.com/eclipse-openj9/openj9/pull/15811#issuecomment-1234797094. I think the openjdk test needs to be updated for virtual thread.

**Comment by hangshao0** [comment]
@tajila Do you have more review comments ?

File: runtime/jvmti/jvmtiHelpers.c [hunk 4]
```diff
@@ -772,6 +772,89 @@ getThreadState(J9VMThread *currentThread, j9object_t threadObject)
 	return state;
 }
 
+#if JAVA_SPEC_VERSION >= 19
+jint
+getVirtualThreadState(J9VMThread *currentThread, jthread thread, BOOLEAN hasVMAccess)
```
Comment (tajila) [inline runtime/jvmti/jvmtiHelpers.c:777]: I think the caller must always have VM access

Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:777]: we should assert this

Comment (hangshao0) [reply runtime/jvmti/jvmtiHelpers.c]: Fixed.

**Comment by fengxue-IS** [comment]
@babsingh can this be merged?

**Comment by babsingh** [comment]
re https://github.com/eclipse-openj9/openj9/pull/15811#issuecomment-1234823232: Which OpenJDK tests were mentioned in this comment?

Do the following JVMTI tests pass with these changes: https://github.com/ibmruntimes/openj9-openjdk-jdk19/tree/openj9/test/hotspot/jtreg/serviceability/jvmti/thread/GetThreadState?

**Comment by hangshao0** [comment]
> Which OpenJDK tests were mentioned in this comment?

It is https://github.com/ibmruntimes/openj9-openjdk-jdk19/tree/openj9/test/hotspot/jtreg/serviceability/jvmti/thread/GetThreadState/thrstat03

We hit the assertion that is recently added. 

`j9jvmti.422    *   ** ASSERTION FAILED ** at /openj9-openjdk-jdk/openj9/runtime/jvmti/jvmtiHelpers.c:177: (((((void *)0) != targetThread) || isVirtualThread))`

After removing this assertion, I get the error that is mentioned in https://github.com/eclipse-openj9/openj9/pull/15811#issuecomment-1234797094

```
>>> thread state: <none> (0)
>>> thread state:  ALIVE RUNNABLE (5)
>>> thread state:  ALIVE WAITING WAITING_WITH_TIMEOUT SLEEPING (225)
>>> thread state:  TERMINATED (2)
>>> thread state: <none> (0)
>>> thread state:  ALIVE RUNNABLE (5)
>>> thread state:  ALIVE WAITING WAITING_INDEFINITELY (145)
Wrong state:  ALIVE WAITING WAITING_INDEFINITELY (145)
   expected:  SLEEPING (64)
>>> thread state:  TERMINATED (2)
```

**Comment by babsingh** [comment]
> @babsingh can this be merged?

Will complete the review by EOD.

**Comment by babsingh** [comment]
re https://github.com/eclipse-openj9/openj9/pull/15811#issuecomment-1245552277:

- How does the RI behave for [GetThreadState/thrstat03](https://github.com/ibmruntimes/openj9-openjdk-jdk19/tree/openj9/test/hotspot/jtreg/serviceability/jvmti/thread/GetThreadState/thrstat03)?
- Only two JVMTI functions allow `targetThread == NULL` when `mustBeAlive == FALSE` for virtual threads: `jvmtiGetThreadState` and `jvmtiGetThreadInfo`. That assertion needs to wrapped with:
```
#if JAVA_SPEC_VERSION >= 19
	if (mustBeAlive) { 
		Assert_JVMTI_true((NULL != targetThread) || isVirtualThread);
	}
#endif /* JAVA_SPEC_VERSION >= 19 */
```

**Comment by hangshao0** [comment]
RI failed the same way as OpenJ9. I believe the test needs to be updated.

```
>>> thread state: <none> (0)
>>> thread state:  ALIVE RUNNABLE (5)
>>> thread state:  ALIVE WAITING WAITING_WITH_TIMEOUT SLEEPING (225)
>>> thread state:  TERMINATED (2)
>>> thread state: <none> (0)
>>> thread state:  ALIVE RUNNABLE (5)
>>> thread state:  ALIVE WAITING WAITING_INDEFINITELY PARKED (657)
Wrong state:  ALIVE WAITING WAITING_INDEFINITELY PARKED (657)
   expected:  SLEEPING (64)
>>> thread state:  TERMINATED (2)
```

**Comment by babsingh** [comment]
re https://github.com/eclipse-openj9/openj9/pull/15811#issuecomment-1245615810:
- RI's output is `ALIVE WAITING WAITING_INDEFINITELY PARKED`. OpenJ9's output is `ALIVE WAITING WAITING_INDEFINITELY`. Why doesn't OpenJ9 set the `PARKED` flag as the RI? Is it because the RI uses `LockSupport.park` for `Object.wait`, and OpenJ9 doesn't?

re https://github.com/eclipse-openj9/openj9/pull/15811#issuecomment-1245615810: 
- Can you also update the assert?

**Comment by hangshao0** [comment]
> RI's output is ALIVE WAITING WAITING_INDEFINITELY PARKED. OpenJ9's output is ALIVE WAITING WAITING_INDEFINITELY. Why doesn't OpenJ9 set the PARKED flag as the RI?

RI's native(JVMTI) API behaviour is inconsistent with its Java API on this. There is no clear documentation on how to translate virtual thread state into `JVMTI_THREAD_STATE_XXX`. There is only documentation to translate `JVMTI_JAVA_LANG_THREAD_STATE_XXX` to `JVMTI_THREAD_STATE_XXX`. The implementation here is according to the Java API `j.l.VirtualThread.threadState()`. I can add JVMTI_THREAD_STATE_PARKED if you want, but this brings the inconsistency into OpenJ9.

**Comment by hangshao0** [comment]
Assertion fixed.

**Comment by babsingh** [comment]
re https://github.com/eclipse-openj9/openj9/pull/15811#issuecomment-1245642969:

I read the test incorrectly. The virtual thread, on which `GetThreadState` is invoked, is actually executing `Thread.sleep`: [thrstat03.java#L102](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/f85c2db894523ca10a10aa9b2a434439c045d217/test/hotspot/jtreg/serviceability/jvmti/thread/GetThreadState/thrstat03/thrstat03.java#L102). As per the [JVMTI doc](https://download.java.net/java/early_access/jdk19/docs/specs/jvmti.html#GetThreadState), a virtual thread in this state can either have `JVMTI_THREAD_STATE_SLEEPING` set or `JVMTI_THREAD_STATE_PARKED` set instead of `JVMTI_THREAD_STATE_SLEEPING`. So, we should match the RI.

**Comment by babsingh** [comment]
re https://github.com/eclipse-openj9/openj9/pull/15811#issuecomment-1245458048: Previously, I had only asked if OpenJ9 passes the following tests: [jvmti/thread/GetThreadState](https://github.com/ibmruntimes/openj9-openjdk-jdk19/tree/openj9/test/hotspot/jtreg/serviceability/jvmti/thread/GetThreadState). Are there any other noticeable differences between RI and OpenJ9 for these tests?

**Comment by hangshao0** [comment]
Updated to have `JVMTI_THREAD_STATE_PARKED` set to match RI.

> Are there any other noticeable differences in thread states between RI and OpenJ9 for these tests?

No.

File: runtime/oti/jvmtiInternal.h [hunk 5]
```diff
@@ -570,7 +570,17 @@ typedef struct jvmtiGcp_translation {
...
+#define JVMTI_VTHREAD_STATE_RUNNABLE 2
+#define JVMTI_VTHREAD_STATE_RUNNING 3
+#define JVMTI_VTHREAD_STATE_PARKING 4
+#define JVMTI_VTHREAD_STATE_PARKED 5
+#define JVMTI_VTHREAD_STATE_PINNED 6
+#define JVMTI_VTHREAD_STATE_YIELDING 7
 #define JVMTI_VTHREAD_STATE_TERMINATED 99
+#define JVMTI_VTHREAD_STATE_SUSPENDED (1<<8)
+#define JVMTI_VTHREAD_STATE_RUNNABLE_SUSPENDED (JVMTI_VTHREAD_STATE_RUNNABLE|JVMTI_VTHREAD_STATE_SUSPENDED)
+#define JVMTI_VTHREAD_STATE_PARKED_SUSPENDED (JVMTI_VTHREAD_STATE_PARKED|JVMTI_VTHREAD_STATE_SUSPENDED)
```
Comment (babsingh) [inline runtime/oti/jvmtiInternal.h:583]: formatting nitpick: missing space.
```suggestion
#define JVMTI_VTHREAD_STATE_SUSPENDED (1 << 8)
#define JVMTI_VTHREAD_STATE_RUNNABLE_SUSPENDED (JVMTI_VTHREAD_STATE_RUNNABLE | JVMTI_VTHREAD_STATE_SUSPENDED)
#define JVMTI_VTHREAD_STATE_PARKED_SUSPENDED (JVMTI_VTHREAD_STATE_PARKED | JVMTI_VTHREAD_STATE_SUSPENDED)
```

File: test/functional/Java19andUp/src/org/openj9/test/jep425/VirtualThreadTests.java [hunk 6]
```diff
@@ -150,7 +150,17 @@ private static int readVirtualThreadStates(Class vthreadCls, String fieldName) t
...
 		final int JVMTI_VTHREAD_STATE_NEW = 0;
+		final int JVMTI_VTHREAD_STATE_STARTED = 1;
+		final int JVMTI_VTHREAD_STATE_RUNNABLE= 2;
+		final int JVMTI_VTHREAD_STATE_RUNNING = 3;
+		final int JVMTI_VTHREAD_STATE_PARKING = 4;
+		final int JVMTI_VTHREAD_STATE_PARKED = 5;
+		final int JVMTI_VTHREAD_STATE_PINNED = 6;
+		final int JVMTI_VTHREAD_STATE_YIELDING = 7;
 		final int JVMTI_VTHREAD_STATE_TERMINATED = 99;
+		final int JVMTI_VTHREAD_STATE_SUSPENDED = (1<<8);
```
Comment (babsingh) [inline test/functional/Java19andUp/src/org/openj9/test/jep425/VirtualThreadTests.java:161]: formatting nitpick: missing space.
```suggestion
		final int JVMTI_VTHREAD_STATE_SUSPENDED = (1 << 8);
```

File: runtime/jvmti/jvmtiThread.c [hunk 7]
```diff
@@ -67,51 +67,61 @@ jvmtiGetThreadState(jvmtiEnv* env,
...
+			/* If thread is NULL, the current thread is used which cannot be a virtual thread.
+			 * There is an assertion inside getVirtualThreadState() that thread is not NULL.
 			 */
-			rc = getVMThread(currentThread, thread, &targetThread, TRUE, FALSE);
-			targetThread = NULL;
-			threadStartedFlag = JNI_FALSE;
-		}
+			rv_thread_state = getVirtualThreadState(currentThread, thread);
+		} else {
+#endif /* JAVA_SPEC_VERSION >= 19 */
```
Comment (babsingh) [inline runtime/jvmti/jvmtiThread.c:77]: With this change, the `ifdef` in the corresponding `}` can be avoided.
```suggestion
		} else
#endif /* JAVA_SPEC_VERSION >= 19 */
		{
```

File: runtime/jvmti/jvmtiThread.c [hunk 8]
```diff
@@ -67,51 +67,61 @@ jvmtiGetThreadState(jvmtiEnv* env,
...
-				vm->internalVMFunctions->haltThreadForInspection(currentThread, targetThread);
-				/* The vmthread can't be recycled because getVMThread() prevented it from exiting. */
-				rv_thread_state = getThreadState(currentThread, targetThread->threadObject);
-				vm->internalVMFunctions->resumeThreadForInspection(currentThread, targetThread);
+				releaseVMThread(currentThread, targetThread, thread);
 			}
-			releaseVMThread(currentThread, targetThread, thread);
+#if JAVA_SPEC_VERSION >= 19
 		}
+#endif /* JAVA_SPEC_VERSION >= 19 */
```
Comment (babsingh) [inline runtime/jvmti/jvmtiThread.c:124]:
```suggestion
		}
```

File: runtime/jvmti/jvmti_internal.h [hunk 9]
```diff
@@ -1193,6 +1193,16 @@ getObjectSize(J9JavaVM *vm, j9object_t obj);
 jint
 getThreadState(J9VMThread *currentThread, j9object_t threadObject);
 
+#if JAVA_SPEC_VERSION >= 19
+/**
+* Get the state of a virtual thread
+* @param *currentThread The current thread.
+* @param thread The virtual thread whose state is queried.
+* @return jint The virtual thread state .
+*/
```
Comment (babsingh) [inline runtime/jvmti/jvmti_internal.h:1202]: formatting
```suggestion
/**
 * Get the state of a virtual thread.
 * @param currentThread The current thread.
 * @param thread The virtual thread whose state is queried.
 * @return jint The virtual thread state .
 */
```

**Review by babsingh — CHANGES_REQUESTED** [review]
Final formatting nitpicks. Plan to launch PR builds after this.

File: runtime/oti/jvmtiInternal.h (hunk 5 above)
Comment (hangshao0) [reply runtime/oti/jvmtiInternal.h]: Updated.

File: test/functional/Java19andUp/src/org/openj9/test/jep425/VirtualThreadTests.java (hunk 6 above)
Comment (hangshao0) [reply test/functional/Java19andUp/src/org/openj9/test/jep425/VirtualThreadTests.java]: Updated.

File: runtime/jvmti/jvmtiThread.c (hunk 7 above)
Comment (hangshao0) [reply runtime/jvmti/jvmtiThread.c]: Updated.

File: runtime/jvmti/jvmti_internal.h (hunk 9 above)
Comment (hangshao0) [reply runtime/jvmti/jvmti_internal.h]: Updated.

File: runtime/jvmti/jvmtiThread.c (hunk 8 above)
Comment (hangshao0) [reply runtime/jvmti/jvmtiThread.c]: Updated.

**Comment by babsingh** [comment]
jenkins test sanity win,xlinux jdk19

**Comment by babsingh** [comment]
jenkins test sanity amac jdk8

**Comment by hangshao0** [comment]
You may want to re-launch the PR builds.

**Comment by babsingh** [comment]
Cancelling old ones. Will launch new ones soon.

**Comment by babsingh** [comment]
jenkins test sanity win,xlinux jdk19

**Comment by babsingh** [comment]
jenkins test sanity amac jdk11

**Review by babsingh — APPROVED** [review]

**Pull request merged**